datafusion/physical-plan/src/topk/mod.rs

```diff
@@ -121,13 +122,37 @@ pub struct TopK {
     /// Common sort prefix between the input and the sort expressions to allow early exit optimization
     common_sort_prefix: Arc<[PhysicalSortExpr]>,
     /// Filter matching the state of the `TopK` heap used for dynamic filter pushdown
-    filter: Option<Arc<DynamicFilterPhysicalExpr>>,
+    filter: TopKDynamicFilters,
     /// If true, indicates that all rows of subsequent batches are guaranteed
     /// to be greater (by byte order, after row conversion) than the top K,
     /// which means the top K won't change and the computation can be finished early.
     pub(crate) finished: bool,
 }
 
+#[derive(Debug, Clone)]
+pub struct TopKDynamicFilters {
+    /// The current *global* threshold for the dynamic filter.
+    /// This is shared across all partitions and is updated by any of them.
+    /// Stored as row bytes for efficient comparison.
+    threshold_row: Arc<ArcSwapOption<Vec<u8>>>,
+    /// The expression used to evaluate the dynamic filter
+    expr: Arc<DynamicFilterPhysicalExpr>,
```

Comment:
Since the expr should only be updated when the threshold_row is updated, I personally suggest controlling them with the same mutex / arcswap thing

so like

```rust
pub struct TopK {
  ...
    /// locking is done for both fields at once
    filter: ArcSwap<TopKDynamicFilters>,
... 
}

struct TopKDynamicFilters {
    threshold_row: Option<Vec<u8>>,
    /// The expression used to evaluate the dynamic filter
    expr: Arc<DynamicFilterPhysicalExpr>,
}

Reply:
Yep good call! I think otherwise we're bound to run into bugs / synchronization  issues. Did it in bcc0dcdc4.